code/contracts/ONEWallet2.sol

```diff
@@ -0,0 +1,157 @@
+    
+    // Management of daily spendings
+    uint256 dailyLimit; // uint128 is sufficient, but uint256 is more efficient since EVM works with 32-byte words.
+    mapping(uint => uint) dailySpendings; // day => amount ; can be cleaned up once in a while to save storage
+    uint cleanupCheckpoint; // the index of the day since epoch, when the last cleanup was made
+    uint16 constant CLEAN_UP_DAILY_SPENDING_BATCH = 1024; // this is to avoid out-of-gas exceptions
+    uint8 AUTO_CLEANUP_DAYS = 100; // do the auto cleanup of daily spending data after every 100 days
+    
+    uint lastExecutedIdx; // stores the index of OTP that was executed last (this is to avoid reply attacks within the same OTP interval by not allowing commits <= than this number)
+
+    enum OperType { TRANSFER, RECOVERY, SET_LAST_RESORT_ADDR }
+
+    struct Commit {
+        bytes32 hash;          // h(eOTP || operType || params)  
+        uint timestamp;      // OTP must be valid in this time; not in the time of reveal        
+        OperType operType;
+        // below are operation-specific parameters, based on operType
+        address payable addr; 
+        uint amount;        
+    }    
+    mapping(uint32 => Commit[]) public commits; // idx of OTP valid during a commitment => commitment data
+
+    event UnknownTransferError(address dest);
+    event UnknownOperationType(OperType otype);
+
+    constructor(bytes32 root_, bytes32 recoveryHash_, uint8 height_, uint8 interval_, uint32 t0_, uint32 lifespan_, address payable lastResortAddress_, uint256 dailyLimit_){        
+        require(lastResortAddress_ != address(0), "Last resort address is not set");
+        root = root_; // Merkle root for OTPs
+        recoveryHash = recoveryHash_;
+        height = height_;
+        interval = interval_;        
+        t0 = t0_; 
+        lifespan_ = lifespan_;
+        lastResortAddress = lastResortAddress_;
+        dailyLimit = dailyLimit_; 
+        cleanupCheckpoint = block.timestamp / 1 days;     // init to the current day
+    }
+
+    receive() external payable {}
+
+    /**
+     * This is a general commit and should works with all kinds of operations distinguished by otype, followed by operation-specific parameters. 
+     */
+    function commit(bytes32 hash, OperType otype, address payable addr, uint amount) external {
+        uint32 submittedOTPIdx = uint32(block.timestamp) / interval - t0;
+        require(submittedOTPIdx > lastExecutedIdx, "It is not possible to submit commits in the current OTP interval anymore. Wait for the next interval.");
+        Commit memory nc = Commit(hash, block.timestamp, otype, addr, amount);
+        commits[submittedOTPIdx].push(nc);                
+    }
+
+    /**
+     * This is a general reveal for all types of operation.
+     */
+    function reveal(bytes32[] calldata neighbors, uint32 otpIdx, bytes32 eotp) external  returns (bool) {                        
```

Comment:
I think this is a nice way of unifying the threes reveal functions into a single one. As mentioned above if we want to hide the nature of the commit, we would then need to add another parameter here (a struct that unifies parameters of all kinds of operations)

Reply:
IMPORTANT: My conclusion is that we have to store parameters of the operation along with the hash of the commit because the smart contract has to verify in reveal() that h(submitted parameters || eotp) == submitted hash. This is how the first matching entry is found. Actually, this approach is the simplified 3-stage protocol for replacing the root hash in SmartOTPs. While I was designing our commit/reveal, I realized that 3-stages can be reduced to 2-stage, where in 1st stage, you submit hash of commit + params.

This has an implication of showing the parameters of the operation, as mentioned above (which should not be an issue).